PWGHF/HFC/Tasks/taskCorrelationDsHadrons.cxx

```diff
@@ -523,15 +522,126 @@ struct HfTaskCorrelationDsHadrons {
   }
   PROCESS_SWITCH(HfTaskCorrelationDsHadrons, processMcRecME, "Process MC Reco ME", false);
 
+  SliceCache cache;
+  Preslice<CandDsMcReco> perCollisionCand = o2::aod::hf_cand::collisionId;
+  Preslice<CandDsMcGen> perCollisionCandMc = o2::aod::mcparticle::mcCollisionId;
+
   /// Ds-Hadron correlation - for calculating candidate reconstruction efficiency using MC reco-level analysis
-  void processMcCandEfficiency(soa::Join<aod::Collisions, aod::FT0Mults, aod::EvSels> const&,
+  void processMcCandEfficiency(soa::Join<aod::Collisions, aod::FT0Mults, aod::EvSels, aod::McCollisionLabels> const& collisions,
                                soa::Join<aod::McCollisions, aod::MultsExtraMC> const&,
                                CandDsMcGen const& mcParticles,
                                CandDsMcReco const& candidates,
                                aod::TracksWMc const&)
   {
     auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
 
+    /// loop over reconstructed collisions
+    for (const auto& collision : collisions) {
```

Comment:
I am not sure about your logic; if you have two reconstructed collisions that point to the same generated collision (split vertex), then you consider twice the same generated particles. Is this really what you want to do? Wouldn't it be better to loop over mcCollisions and check if they have at least a reconstructed collision associated to it?